File: web/api/routers/labs.py
from __future__ import annotations

from fastapi import APIRouter, HTTPException

from ..models import LabSubmitRequest, LabSubmitResponse

router = APIRouter()

LAB_ANSWERS: dict[str, dict[str, str]] = {
    "caesar-fundamentals": {
        "expected": "KHOOR",
        "hint": "Try shifting each letter back by the key amount.",
        "feedback_correct": "Correct! Caesar cipher with shift 3 transforms HELLO to KHOOR.",
        "feedback_incorrect": "Not quite. Remember: each letter shifts by the key value.",
    },
    "vigenere-breaking": {
        "expected": "KEY",
        "hint": "Use Kasiski examination to find repeated patterns, then determine key length.",
        "feedback_correct": "Correct! The Vigenère key is KEY.",
        "feedback_incorrect": "Try using Kasiski examination on the ciphertext first.",
    },
    "frequency-analysis": {
        "expected": "monoalphabetic",
        "hint": "Check if letter frequencies match English distribution (single peaks).",
        "feedback_correct": "Correct! This text uses monoalphabetic substitution.",
        "feedback_incorrect": "Think about whether the frequency distribution shows one alphabet or multiple.",
    },
    "playfair-cracking": {
        "expected": "SECRET",
        "hint": "Start with common digraphs like TH, HE, IN. Try crib-dragging.",
        "feedback_correct": "Correct! The Playfair key is SECRET.",
        "feedback_incorrect": "Try crib-dragging with common English digraphs.",
    },
    "field-ciphers": {
        "expected": "ATTACK",
        "hint": "First identify the ADFGVX substitution, then reverse the columnar transposition.",
        "feedback_correct": "Correct! You've broken the ADFGVX cipher.",
        "feedback_incorrect": "Remember: ADFGVX uses a 6x6 Polybius square followed by columnar transposition.",
    },
}


@router.post("/labs/{lab_id}/submit", response_model=LabSubmitResponse)
async def submit_lab(lab_id: str, req: LabSubmitRequest) -> LabSubmitResponse:
    if lab_id not in LAB_ANSWERS:
        raise HTTPException(
            status_code=404,
            detail=f"Lab '{lab_id}' not found. Available: {', '.join(sorted(LAB_ANSWERS))}",
        )

    lab = LAB_ANSWERS[lab_id]
    correct = req.answer.strip().upper() == lab["expected"].upper()

    if correct:
        return LabSubmitResponse(
            correct=True,
            feedback=lab["feedback_correct"],
        )
    else:
        return LabSubmitResponse(
            correct=False,
            feedback=lab["feedback_incorrect"],
            hint=lab["hint"],
        )

File: web/static/js/labs.js
(function() {
    const container = document.getElementById('labs-container');

    const labs = [
        {
            id: 'caesar-fundamentals',
            title: 'Lab 1: Caesar Cipher Fundamentals',
            difficulty: 'beginner',
            description: 'Encrypt the word "HELLO" using the Caesar cipher with shift 3. What is the ciphertext?',
            hint: 'Each letter shifts forward by the key value in the alphabet.',
        },
        {
            id: 'vigenere-breaking',
            title: 'Lab 2: Breaking the Vigenère Cipher',
            difficulty: 'intermediate',
            description: 'Use Kasiski examination to determine the keyword used to encrypt a Vigenère ciphertext. The key is a common English word.',
            hint: 'Look for repeated trigrams in the ciphertext. The distance between repeats reveals the key length.',
        },
        {
            id: 'frequency-analysis',
            title: 'Lab 3: Frequency Analysis',
            difficulty: 'intermediate',
            description: 'Analyze the following ciphertext and determine what type of cipher was used. The answer should be one of: monoalphabetic, polyalphabetic, or transposition.',
            hint: 'Compare the letter frequency distribution to standard English. Single-alphabet substitution preserves frequency peaks.',
        },
        {
            id: 'playfair-cracking',
            title: 'Lab 4: Breaking the Playfair Cipher',
            difficulty: 'advanced',
            description: 'Use crib-dragging with common English digraphs (TH, HE, IN, ER) to recover the Playfair keyword. The keyword is a single English word.',
            hint: 'Start with the most common English digraph "TH" and try inserting it at different positions.',
        },
        {
            id: 'field-ciphers',
            title: 'Lab 5: WWI Field Cipher (ADFGVX)',
            difficulty: 'doctoral',
            description: 'Break an ADFGVX cipher by first reversing the columnar transposition, then decoding the ADFGVX substitution. What was the original plaintext message?',
            hint: 'The ADFGVX cipher has two stages: (1) columnar transposition, (2) 6x6 Polybius square lookup.',
        },
    ];

    function renderLabs() {
        container.innerHTML = '';
        labs.forEach(lab => {
            const card = document.createElement('div');
            card.className = 'lab-card';
            card.innerHTML = `
                <h3>${lab.title}</h3>
                <span class="difficulty ${lab.difficulty}">${lab.difficulty}</span>
                <p class="description">${lab.description}</p>
                <div class="form-row">
                    <label for="lab-answer-${lab.id}">Your Answer</label>
                    <input type="text" id="lab-answer-${lab.id}" placeholder="Enter your answer">
                </div>
                <button class="btn-primary lab-submit" data-lab="${lab.id}" type="button">Submit Answer</button>
                <div id="lab-result-${lab.id}" class="hidden"></div>
                <p class="hint" id="lab-hint-${lab.id}" class="hidden"></p>
            `;
            container.appendChild(card);
        });

        container.querySelectorAll('.lab-submit').forEach(btn => {
            btn.addEventListener('click', async () => {
                const labId = btn.dataset.lab;
                const input = document.getElementById(`lab-answer-${labId}`);
                const resultDiv = document.getElementById(`lab-result-${labId}`);
                const hintP = document.getElementById(`lab-hint-${labId}`);
                const answer = input.value.trim();

                if (!answer) { alert('Please enter an answer.'); return; }

                btn.disabled = true;
                btn.textContent = 'Checking...';

                try {
                    const result = await api(`/api/labs/${labId}/submit`, {
                        method: 'POST',
                        body: JSON.stringify({ answer, lab_id: labId }),
                    });

                    resultDiv.className = `lab-result ${result.correct ? 'correct' : 'incorrect'}`;
                    resultDiv.textContent = result.feedback;
                    resultDiv.classList.remove('hidden');

                    if (result.correct) {
                        hintP.textContent = '';
                        hintP.classList.add('hidden');
                    } else if (result.hint) {
                        hintP.textContent = 'Hint: ' + result.hint;
                        hintP.classList.remove('hidden');
                    }
                } catch (e) {
                    resultDiv.className = 'lab-result incorrect';
                    resultDiv.textContent = 'Error: ' + e.message;
                    resultDiv.classList.remove('hidden');
                } finally {
                    btn.disabled = false;
                    btn.textContent = 'Submit Answer';
                }
            });
        });
    }

    router.register('labs', renderLabs);
})();
